pkg/playback/http_replayer.go

```diff
@@ -0,0 +1,198 @@
+	// the status code to 200 if not already set.
+	copyHTTPHeader(w.Header(), wrappedResponse.Headers) // header map must be written before calling w.WriteHeader
+	w.WriteHeader(wrappedResponse.StatusCode)
+	io.Copy(w, bytes.NewBuffer(wrappedResponse.Body))
+
+	// --- Handle webhooks
+	// Check the cassette to see if there are pending webhooks we should fire
+
+	// Note on calling resp.Body().Close() and avoiding a resource leak: all response bodies will be closed in the goroutine below
+	webhookRequests, webhookResponses, err := httpReplayer.readAnyPendingWebhookRecordingsFromCassette() // nolint
+
+	if err != nil {
+		fmt.Printf("Error when checking cassette for webhooks to replay: %v\n", err)
+	}
+	fmt.Printf("Replaying %d webhooks.\n", len(webhookRequests))
+
+	// Send the webhooks
+
+	// Struct used to parse the event.Type from recorded webhook JSON bodies
+	type stripeEvent struct {
+		Type string `json:"type"`
+	}
+
+	go func() {
+		// Note: if there are any errors in processing recorded webhooks here,
+		// we log the error and keep going.
+		// TODO: is the above the appropriate behavior?
+		httpReplayer.replayLock.Wait()       // only send webhooks after the previous request/response is handled
+		httpReplayer.replayLock.Add(1)       // acquire lock so we can send webhooks
+		defer httpReplayer.replayLock.Done() // release lock when done sending webhooks
+
+		for i, webhookReq := range webhookRequests {
+			resp, err := forwardRequest(webhookReq, httpReplayer.webhookURL)
+			if err != nil {
+				fmt.Printf("ERROR when forwarding webhook requests: %v", err)
+				continue
+			}
+			defer resp.Body.Close()
+
+			expectedResp := webhookResponses[i]
+			var evt stripeEvent
+
+			if err != nil {
+				fmt.Printf("ERROR when forwarding webhook requests: %v", err)
+				continue
+			}
+			err = json.Unmarshal(webhookReq.Body, &evt)
+			if err != nil {
+				fmt.Printf("ERROR when forwarding webhook requests: %v", err)
+				continue
+			}
+
+			fmt.Printf("	> Forwarding webhook [%v].\n", evt.Type)
+			fmt.Printf("	> Received %v from client. Expected %v.\n\n", resp.StatusCode, expectedResp.StatusCode)
+			if err != nil {
+				fmt.Printf("ERROR when forwarding webhook requests: %v", err)
+				continue
+			}
+		}
+	}()
+}
+
+// returns error if something doesn't match the cassette
+func (httpReplayer *replayServer) getNextRecordedCassetteResponse(request *httpRequest) (response *httpResponse, err error) {
+	// the passed in request arg may not be necessary
+	uncastResponse, err := httpReplayer.replayer.write(request)
+	if err != nil {
+		return &httpResponse{}, err
+	}
+
+	wrappedResponse := (*uncastResponse).(httpResponse)
+
+	return &wrappedResponse, err
+}
+
+// Reads any contiguous set of webhook recordings from the start of the cassette
+// The caller must call response.Body().Close() on each of the returned http.Response's, otherwise there will be a resource leak
```

Comment:
This comment is out of date, right? Since we are using the "wrapped" httpResponse, instead of http.Response directly?

Reply:
Thanks for catching - fixed in 2f60063d2bd1e627c2288aadfe50345bdf191a1e.